framework/src/kernels/CoupledForce.C

```diff
@@ -27,30 +27,36 @@ CoupledForce::validParams()
   return params;
 }
 
-CoupledForce::CoupledForce(const InputParameters & parameters)
-  : Kernel(parameters), _v_var(coupled("v")), _v(coupledValue("v")), _coef(getParam<Real>("coef"))
+template <bool is_ad>
+CoupledForceTempl<is_ad>::CoupledForceTempl(const InputParameters & parameters)
+  : GenericKernel<is_ad>(parameters),
+    _v_var(coupled("v")),
+    _v(this->template coupledGenericValue<is_ad>("v")),
+    _coef(this->template getParam<Real>("coef"))
 {
   if (_var.number() == _v_var)
-    mooseError("Coupled variable 'v' needs to be different from 'variable' with CoupledForce, "
-               "consider using the CoefReaction kernel or something similar");
+    mooseError("Coupled variable 'v' needs to be different from 'variable' with CoupledForce / "
+               "ADCoupledForce, consider using the CoefReaction kernel or something similar");
 }
 
-Real
-CoupledForce::computeQpResidual()
+template <bool is_ad>
+GenericReal<is_ad>
+CoupledForceTempl<is_ad>::computeQpResidual()
 {
   return -_coef * _v[_qp] * _test[_i][_qp];
 }
 
+template <bool is_ad>
 Real
-CoupledForce::computeQpJacobian()
-{
-  return 0;
-}
-
-Real
-CoupledForce::computeQpOffDiagJacobian(unsigned int jvar)
+CoupledForceTempl<is_ad>::computeQpOffDiagJacobian(unsigned int jvar)
 {
+  // This function will never be called for the AD version. But because C++ does
+  // not support an optional function declaration based on a template parameter,
+  // we must keep this this template for all cases.
```

Comment:
```suggestion
  // we must keep this template for all cases.
```

Reply:
mooseAssert(!is_ad, "...") could be an option. Your call

Reply:
Actually kind of like the idea of using an assert but also keeping the comment around. The assert tips one off that something is very wrong with computeJacobian, but the comment might give some useful extended context. I'm also generally for more comments rather than not enough :)